products/llbuild-analyze/Sources/CriticalPathTool.swift

```diff
@@ -0,0 +1,274 @@
+            to: { $0.pretty = $1 })
+        
+        binder.bind(
+            option: parser.add(option: "--quiet", shortName: "-q", usage: "There will be no output on stdout."),
+            to: { $0.quiet = $1 })
+        
+    }
+    
+    public override func runImpl() throws {
+        guard let databasePath = options.database else {
+            throw ArgumentParserError.expectedValue(option: "database path")
+        }
+        
+        let database = try BuildDB(path: databasePath.pathString, clientSchemaVersion: UInt32(options.clientSchemaVersion))
+        let allKeysWithResult = try database.getKeysWithResult()
+        let solver = CriticalBuildPath.Solver(keys: allKeysWithResult)
+        let path = solver.run()
+        
+        // Output
+        if let outputPath = options.outputPath {
+            let data: Data
+            switch options.outputFormat {
+            case .json:
+                data = try json(path, allKeyResults: allKeysWithResult, buildKeyLookup: solver.keyLookup)
+            case .graphviz:
+                data = graphViz(path, buildKeyLookup: solver.keyLookup)
+            }
+            try verifyOutputPath()
+            FileManager.default.createFile(atPath: outputPath.pathString, contents: data)
+        }
+        
+        if options.quiet { return }
+        print(path.elements.isEmpty ? "Couldn't critical path from database at \(options.database.pathString) because no builds were build." : "Critical Path:\n\(path)")
+    }
+    
+    public func json(_ path: CriticalBuildPath, allKeyResults: BuildDBKeysWithResult, buildKeyLookup: IdentifierFactory<BuildKey>) throws -> Data {
+        let encoder = JSONEncoder()
+        if options.pretty {
+            encoder.outputFormatting = [.prettyPrinted]
+            if #available(OSX 10.13, *) {
+                encoder.outputFormatting.insert(.sortedKeys)
+            }
+        }
+        return try encoder.encode(PathOutput(path, allKeyResults: allKeyResults, buildKeyLookup: buildKeyLookup))
+    }
+    
+    private func verifyOutputPath() throws {
+        guard let outputPath = options.outputPath else { return }
+        if FileManager.default.fileExists(atPath: outputPath.pathString) {
+            throw StringError("Can't output critical path to \(outputPath) - file exists.")
+        }
+    }
+    
+    private func graphViz(_ path: CriticalBuildPath, buildKeyLookup: IdentifierFactory<BuildKey>) -> Data {
+        var result = "digraph G {\n\tedge [style=dotted]\n"
+        
+        func string(for key: BuildKey) -> String {
+            // TODO: This is obviously just temporary - need a shorter version for visualization
+            return buildKeyLookup.identifier(element: key).description
+        }
+        
+        for element in path {
+            for dep in element.result.dependencies {
+                result += "\t\"" + string(for: dep) + "\" -> \"" + string(for: element.key) + "\"\n"
```

Comment:
nit: using string interpolation would probably be more readable here.

Reply:
Yes, a bit more 👍